flutter/lib/src/span_frame_metrics_collector.dart

```diff
@@ -0,0 +1,201 @@
+import 'dart:collection';
+
+import 'package:flutter/cupertino.dart';
+import 'package:meta/meta.dart';
+
+import '../sentry_flutter.dart';
+
+import 'frame_callback_handler.dart';
+import 'native/sentry_native.dart';
+
+@internal
+class SpanFrameMetricsCollector implements PerformanceContinuousCollector {
+  final SentryFlutterOptions options;
+  final FrameCallbackHandler? _frameCallbackHandler;
+  final SentryNative? _native;
+
+  final frames = SplayTreeMap<DateTime, int>();
+  final runningSpans = <ISentrySpan>[];
+
+  bool get isFrameTrackingPaused => _isFrameTrackingPaused;
+  bool _isFrameTrackingPaused = true;
+
+  bool get isFrameTrackingRegistered => _isFrameTrackingRegistered;
+  bool _isFrameTrackingRegistered = false;
+
+  final _stopwatch = Stopwatch();
+
+  SpanFrameMetricsCollector(this.options,
+      {FrameCallbackHandler? frameCallbackHandler, SentryNative? native})
+      : _frameCallbackHandler =
+            frameCallbackHandler ?? DefaultFrameCallbackHandler(),
+        _native = native ?? SentryFlutter.native;
+
+  @override
+  void onSpanStarted(ISentrySpan span) {
+    if (span is NoOpSentrySpan || !options.enableFramesTracking) {
+      return;
+    }
+
+    runningSpans.add(span);
+    startFrameTracking();
+  }
+
+  @override
+  Future<void> onSpanFinished(ISentrySpan span, DateTime endTimestamp) async {
+    if (span is NoOpSentrySpan || !options.enableFramesTracking) {
+      return Future.value();
+    }
+
+    await recordSpanFrameMetrics(span, endTimestamp);
+
+    if (runningSpans.isEmpty) {
+      clear();
+    } else {
+      final oldestSpan = runningSpans.first;
+      frames.removeWhere((key, value) {
+        return key.isBefore(oldestSpan.startTimestamp);
+      });
+    }
+  }
+
+  /// Calls [WidgetsBinding.instance.addPersistentFrameCallback] which cannot be unregistered
+  /// and exists for the duration of the application's lifetime.
+  ///
+  /// Stopping the frame tracking means setting `isFrameTrackingPaused = true`
+  /// to prevent actions being done when the frame callback is triggered.
+  void startFrameTracking() {
+    _isFrameTrackingPaused = false;
+
+    if (!_isFrameTrackingRegistered) {
+      _frameCallbackHandler?.addPersistentFrameCallback(frameCallback);
+      _isFrameTrackingRegistered = true;
+    }
+  }
+
+  void frameCallback(Duration duration) async {
+    if (_isFrameTrackingPaused) {
+      return;
+    }
+
+    if (_stopwatch.elapsedMilliseconds == 0) {
+      _stopwatch.start();
+    }
+
+    await _frameCallbackHandler?.endOfFrame;
+    _stopwatch.stop();
+
+    // ignore: invalid_use_of_internal_member
+    frames[getUtcDateTime()] = _stopwatch.elapsedMilliseconds;
+
+    _stopwatch.reset();
+    if (_frameCallbackHandler?.hasScheduledFrame == true) {
+      _stopwatch.start();
+    }
+  }
+
+  Future<void> recordSpanFrameMetrics(
+      ISentrySpan span, DateTime endTimestamp) async {
+    final displayRefreshRate = await _native?.displayRefreshRate();
```

Comment:
I'm not sure whether we always need to fetch the refresh rates

my thinking was it may update dynamically but is that really the case for ios and android?

Reply:
In theory yes, there are variable refresh rates devices on both Android and iOS
> or maybe we should fetch it in onSpanStarted

yeah, maybe it's better, so we don't await for it here.
Also i would fallback to 60 hz, instead of dropping the frames